### PR TITLE
Grouped bar chart

### PR DESCRIPTION
**Pull request by gazal-k**
Modification of https://github.com/dc-js/dc.js/pull/679
- Hopefully I have resolved most of the conflicts with dc.js mainline.
- Resolved a few issue in rendering multi/grouped bar chart.
- I've added some basic tests for multi/grouped bar chart.

Pending:
- not sure what functionalities expected of bar charts in general is missing in the multi/grouped version.
- multi/grouped row chart.

Please feel free to suggest changes or fork this and fix further.

### CODE REVIEW

File: src/multigroup-mixin.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        _chart._groupGap = _;
+        return _chart;
+    };
+
+    _chart.legendables = function () {
+        var items = [];
+        _chart.groups().forEach(function (g, k) {
+            var item = {
+                chart:  _chart,
+                name:   g.title,
```
Comment (gazal-k) [inline src/multigroup-mixin.js]: this approach doesn't seem consistent. the legend name for the components will appear only if each group has a property 'title'

File: src/multibar-chart.js [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                    x += _gap / 2;
+                    return dc.utils.safeNumber(x);
+                })
+                .attr('y', function (d) {
+                    return _chart.y()(0);
+                });
+
+        /*if (_chart.renderTitle()) {
+            bars.append('title').text(dc.pluck('data', _chart.title(d.name)));
+        }*/
```
Comment (gazal-k) [inline src/multibar-chart.js]: not sure how to render titles

**Comment by gordonwoodhull** [comment]
Oh wow, now we have two competing versions of grouped bars. See #984. This is going to get interesting.

The issue I have with this one, if I remember correctly, is that @wssbck made a ton of unrelated changes in this branch. Enough so that at the time, I couldn't make heads or tails of it. 

However, this is less of an issue now that you've fixed the tests, @gazal-k. Thank you! Since you've reviewed it, can you summarize what the other changes are?

**Comment by gazal-k** [comment]
@gordonwoodhull I didn't really know about #984. I originally had a hack which involved some transform = translate() style post render tricks to get multi/grouped bar charts functionality within composite chart. Then I saw the discussion between you and @wssbck about how its probably not the best idea to do grouped bar within composite.

**Comment by gazal-k** [comment]
I also had a hard time figuring out all of the code here. I don't have a comprehensive idea of how the framework gets things done. I just meddled with it.

This is what I understand about it:
 - There's a multigroup-mixin which abstracts some functionality for grouped bar charts and row charts.
 - The grouped bar chart works for the most part after some tweaking; I was able to write some basic tests.
 - Yet to check the grouped row chart

If you can please check some inline notes I have added and help me, that'd be really awesome.

**Comment by gazal-k** [comment]
@gordonwoodhull I did take some inspiration from #984 to make the code cleaner.

With the current approach I was able to fix title rendering, and labeling groups became more consistent.
I liked how @alan-unravel was able to get the functionality within dc.barChart.

**Comment by gazal-k** [comment]
@gordonwoodhull I think I finally figured out the heads and tails of it :smile: 

Got the grouped bar chart functionality to work within bar chart code.

File: src/bar-chart.js [hunk 3]
```diff
@@ -246,6 +265,21 @@ dc.barChart = function (parent, chartGroup) {
...
+    dc.override(_chart, 'data', function () {
+        var data = _chart._data();
+
+        if (_groupBars) {
+            // set y0 to 0 on all layers
+            data.forEach(function (layer) {
+                layer.values.forEach(function (value) {
+                    value.y0 = 0;
+                });
+            });
```
Comment (gazal-k) [inline src/bar-chart.js]: not sure whether this is the best approach to get this done

**Comment by gazal-k** [comment]
Fixes #558

**Comment by gazal-k** [comment]
got to get centerBar to work properly when groupBars is set.

Another thing I noticed, with ordinal x axis, the bars seem to already be centered (with single group, stacks both vertical and horizontal) and applying centerBar(true) seems to break it. How should this be ideally?

File: src/bar-chart.js [hunk 4]
```diff
@@ -147,8 +160,17 @@ dc.barChart = function (parent, chartGroup) {
...
             .attr('x', function (d) {
                 var x = _chart.x()(d.x);
-                if (_centerBar) {
-                    x -= _barWidth / 2;
+                if (_groupBars) {
+                    x += _chart.groupGap() / 2;
+                    x += layerIndex * (_barWidth + _gap);
+                    x += _gap / 2;
+                }
+                if (_centerBar && !_chart.isOrdinal()) {
```
Comment (gazal-k) [inline src/bar-chart.js]: this check; _chart.isOrdinal() affects bar charts in general.

I saw that bar charts with ordinal x axis are already centered around axis points. So setting centerBar(true) only seems to offset it incorrectly.

Please see if this is fine

**Comment by madslundgaard** [comment]
Dear @gazal-k and @gordonwoodhull,
How far would You think this is from being submitted to development branch. I am working on a project where grouped bar chart is a critical chart type (for version/scenario comparison).

I would like to help on the issue if you believe that additional input/testing is needed (basic JS,D3,DC experience).

**Comment by gazal-k** [comment]
@madslundgaard you needn't wait for this PR to get merged if u want to use this immediately. U can create a fork of dcjs and merge my branch to yours and use that instead of referring to dcjs base fork.

If you are using bower, the dependency can be something like: `madslundgaard/dc.js#develop`

**Comment by gordonwoodhull** [comment]
Yes, @madslundgaard, in fact testing the feature and providing feedback, and code review if you're up for it, is the best way to help it get into dc.js. Adding more tests is always welcome too!

Note that @mtraynham and others are discussing adding another mixin to provide a uniform interface for "extra dimensions" like this one. I don't think it will change the external interface too much, just help with the code organization.

**Comment by alsguimaraes** [comment]
I second  @madslundgaard 's question.  Would you know which of the two grouped bar chart implementations will be added to 2.1? I am willing to test it too. It will take a while, as I need to setup bower, etc.

**Comment by gordonwoodhull** [comment]
I think this is the more likely one to merge because it has tests. However it will still take some careful review. Trying it out and reporting your experience here is a big help.

**Comment by Sid0403** [comment]
Hi, I want to create a grouped bar chart  and came across this thread.
I am very new to the entire D3, DC setup. How can I use this PR to create a grouped bar chart? I could really use some help here .. Thanks in advance!

**Comment by MsJR** [comment]
I'm beginner in dc.js and I want to draw a grouped bar chart..
I tried to use compositeChart but it didn't work for me..
Here is my problem : 
http://stackoverflow.com/questions/36949955/is-it-possible-to-draw-a-grouped-bar-chart-with-dc-js
I saw the code here, but I don't know how to use it.. can anyone explain or add comments about how to use and run the code.. and thank you

**Comment by gordonwoodhull** [comment]
Hi @Sid0403, @MsJR, I've put @gazal-k's changes on a branch `gazal-grouped-bar`, so you can try it out by just going to the branch here:
https://github.com/dc-js/dc.js/tree/gazal-grouped-bar

Or download the .zip file of this branch here:
https://github.com/dc-js/dc.js/archive/gazal-grouped-bar.zip

There is currently no example but the API looks very simple, just set the [groupBars](https://github.com/dc-js/dc.js/blob/gazal-grouped-bar/web/docs/api-latest.md#dc.barChart+groupBars) flag and the [groupGap](https://github.com/dc-js/dc.js/blob/gazal-grouped-bar/web/docs/api-latest.md#barchartgroupgap--number--barchart) size.

I added a pretty dumb example here:
https://github.com/dc-js/dc.js/blob/gazal-grouped-bar/web/examples/grouped-bar.html

In particular, it seems that it needs `.centerBar(true)` for the ticks to land within the groups.

Please raise any issues or requests on this ticket. Thanks for your interest!

**Comment by slackhideo** [comment]
This versions of grouped bars works well.
Thanks, @gazal-k and @gordonwoodhull for your effort.

I noticed that the ordering and barPadding functions do not work together with groupBars. Is there a way I can have grouped bars working with these two?

Thanks.

**Comment by gordonwoodhull** [comment]
Hi @slackhideo, thanks for your review. Have you also compared #984? 

I'll let @gazal-k speak to whether he has time to add ordering and barPadding support.

**Comment by slackhideo** [comment]
Hi, @gordonwoodhull. Thanks for the reply.

Actually I was wrong about the ordering support. It does support this function (sorry!).
Instead of using `chart.ordering(function(d) { return -d.value; })` I needed to use `chart.ordering(function(d) { return -d.value['1']; })` (to order according to the layer '1').

But I'm afraid I have no idea how to tackle the barPadding support.

**Comment by gordonwoodhull** [comment]
@slackhideo, have you also tried #984?

**Comment by slackhideo** [comment]
Oh, sorry @gordonwoodhull, I forgot that part.
Now I tested #984 and found that it supports barPadding.
Thanks for the suggestion.

I've chosen this implemention because, judging by your example, this version looks better to me. The groups are more aligned with the X-axis' ticks (turning off/on centerBar didn't help here) and the spacing between bars in the same group seems more pleasant. I don't know if these things are easily changeable.
I guess both versions have their strengths and shortcomings.

**Comment by gazal-k** [comment]
@gordonwoodhull @slackhideo sorry, its been quite a while since I've gotten my hands dirty with dc.js. I've been and still am tied up with some other work. I'd love to contribute when I get some free time. But until then, sorry :disappointed:

**Comment by gordonwoodhull** [comment]
Thanks @slackhideo for your review. That's very helpful.

And thanks @gazal-k for all your contributions! Great stuff. 

I think the issue might be that we need two levels of barPadding for this chart. But my memory is a bit fuzzy. I'll try to take a look when I can.

**Comment by slackhideo** [comment]
Thank you, @gordonwoodhull and @gazal-k. Your work is amazing, really.
It's just some minor things that can be improved (well, almost everything can be improved), but overall is a stunning library and this grouped bar feature is very useful.
Your hard work is appreciated!

**Comment by frankisans** [comment]
Hi @gordonwoodhull I have check both branches on a composite char with a line chart and a grouped bar chart. I prefer gazak branch because no gap on same group bars:

gazak:
image

kavanagh:
image

I use another control to change the focus dinamically but the bars overlap the yAxis:
image

Seems not a problem with grouped bar chart implementation because gets the same result with normal stacked bar. I cannot use padding because elasticX is false for focus to work.

**Comment by suntong** [comment]
- All checks have passed
- This branch has no conflicts with the base branch

Waiting for the merge...

**Comment by gordonwoodhull** [comment]
Thanks @suntong, I agree this is very close to being ready to merge (and it will be the first thing to go into 2.1). However, there are a still a few questions raised by the author which I need to address, as well as manual code review and testing.

Please go ahead and use the version in this branch:
https://github.com/dc-js/dc.js/tree/gazal-grouped-bar

I've pulled the latest changes from develop, so this has all of the fixes from dc.js 2.0.

I apologize for the long turn-around but unfortunately this is not my full-time job (however much I wish it were).

**Comment by suntong** [comment]
Definitely, totally understand. I was just trying to say something meaningful in order to subscribe to this thread. Take your time.

**Comment by carbonisle** [comment]
Have a couple of quick questions - and I apologize in advance for my naivety, I'm still learning how to use D3, DC, Crossfilter, et el.  I'm using the sample HTML from above as a baseline.

1) For some reason, I have to use numbers in my data as opposed to strings.  (Convert "male" and "female" to 1/0).  I'm guessing it has to do with the reduce functions I'm using.  This also effects my x-axis labels, of course.  I'd rather they show the text variations.

`    ndx = crossfilter(eData),
            groupDim = ndx.dimension(function(d) {return d.service;}),
            qtySumGroup = groupDim.group().reduce(function(p,v) {
                p[v.gender] = (p[v.gender] || 0) + v.qty;
                return p;
            }, function(p,v) {
                p[v.gender] = (p[v.gender] || 0) - v.qty;
                return p;
            }, function() {
                return{};
            });`

2) I'm also noticing that it doesn't seem to crossfilter the data.  When I click one of the bars in a group, it doesn't filter my other charts on the page.  What am I missing?

Again, I apologize for these, likely obvious and easy, questions - but I'm on a deadline and I'm spinning my tires self-teaching.  Need a little nudge.

**Comment by gordonwoodhull** [comment]
Hi @carbonisle, could you please open a question on Stack Overflow with the [dc.js] tag? I will be sure to answer it quickly, especially if you are able to create a fiddle or point to a web site which demonstrates the problems.

These do not sound like they are issues with dc.js, nor do they sound like they have to do with the grouped bar chart PR. So I'd rather not clutter up this already very long ticket. Thank you!

**Comment by parliament718** [comment]
@gordonwoodhull Could you by any chance merge develop again? 57 commits and almost 3 months behind and Im having issues integrating @types/dc, which follows this develop branch.

**Comment by fatangare** [comment]
Does it work with time scale or only with ordinal?

**Comment by Frozenlock** [comment]
Very interested in using grouped bar chart!

Is there still a strong possibility of it being added soon in the official library? In other words, is it wise to start building on top of this?

**Comment by gordonwoodhull** [comment]
@fatangare, because it build on the stacked chart, it will work wherever that does. So yes.

@Frozenlock, yes, I hope so. This is one of a few major PRs ready to merge before the big port to d3v4.

**Comment by OivalfMarques** [comment]
This solution does not allow having  bar chart multiple series and staked at the same time, but even so this solution moves away from the rest of the library. What I propose is that instead of looking at the number of stacks, start looking at the number of bar chart, if this is a composite chart.

Follow the implementation link on the latest version of this library

https://plnkr.co/edit/V99X09LPMNc0V7qP3jum?p=preview

I'm waiting for an feedback

Thank you for your time.

**Comment by gordonwoodhull** [comment]
Hi @OivalfMarques, do you mean that you have an implementation of stacked *and* grouped charts? That would of course be fantastic.

**Comment by Frozenlock** [comment]
@gordonwoodhull This indeed looks like an implementation of grouped and stacked charts, which is amazing. (You can see that the gap is a little wider between groups.)

@OivalfMarques Amazing work! Do you have a fork available with your changes? PR?

**Comment by OivalfMarques** [comment]
hi @Frozenlock yes i have.

**Comment by gazal-k** [comment]
This is most likely unnecessary at this point.
